Timeseries - Moving Average Detector

viveksjain opened the pull request:
As promised, time series detector along with code to run the Numenta benchmark. And the results are in for moving average! *drumroll*
Final score for 'macrobase' detector on 'ma_reward_low_FP_rate' profile = 0.68
Final score for 'macrobase' detector on 'ma_reward_low_FN_rate' profile = 15.22
Final score for 'macrobase' detector on 'ma_standard' profile = 5.59

Compared to [random](https://github.com/numenta/NAB#scoreboard) Reward Low FP = 5.8, Reward Low FN = 25.9, Standard Profile = 16.8. :)

viveksjain commented:
Also, there's a couple of TODOs still in the code - I'm going to try to fix up most of them in the next couple of weeks.

pbailis commented:
Planning to look this over tonight.

pbailis commented on script/numenta_benchmark.py:
```diff
@@ -0,0 +1,135 @@
+import re
+import json
+
+# Note: initially need to run `python scripts/create_new_detector.py --detector macrobase_ma`
```
What's the long-term gameplan for the `nab` branch?

viveksjain replied:
That's a good question. slf4j would have been ideal since it's already being used, but unfortunately it's buggy. The options I see are: (i) keep the nab branch, updating as necessary. I don't think it's a huge deal if it gets out of date. (ii) Adding another flag. I'm not a fan of having a ton of different flags, but if we want to use this longer term it would probably be the better approach.

pbailis replied:
I'd love to have numenta working long-term. What's buggy about slf4j? I'm happy to migrate if you have a preferred alternative.

pbailis commented on src/main/java/macrobase/analysis/outlier/MovingAverage.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    @Override
+    public double scoreWindow(Queue<Datum> data, Datum newDatum, Datum oldDatum) {
```
So, this is basically a tuple-based sliding window model, where `oldDatum` is what just exited the window. There's another way to do this so you talk about windows composed of `panes`, where each pane can contain multiple transactions and the window slides pane by pane.

viveksjain replied:
Makes sense, but seems like a slightly more complicated API. Is there a model that requires pane support? Otherwise, in my opinion, simplicity > unused generality.

pbailis replied:
The main model I want to support is real-time sliding window queries -- e.g., give me the EWMA over the last five minutes. If you think the tuple-at-a-time interface is a good fit for expressing this, I'm okay with it.

viveksjain replied:
There's a distinction to be made between the interface for algorithm writers and for algorithm users. For algorithm writers, I think it's generally easier to think in terms of tuple-at-a-time, and this interface is designed for them. For consumers of the algorithm's scoring, it may be easier to use the panes idea, but that can be a different interface/class that calls the TimeSeriesOutlierDetector interface inside.

viveksjain replied:
For posterity: Peter and I discussed offline and I realized the issue with this `scoreWindow` is essentially that it assumes the window is of a fixed size, and every time you remove a tuple you also add another one. This may not be the case if the window, for example, represents a fixed time range and the points are not uniformly spread out. This can be fixed by having interfaces that allow individually adding/removing one (or many) tuples at a time.

pbailis commented on src/main/java/macrobase/ingest/SQLLoader.java:
```diff
@@ -88,6 +93,7 @@ public RowSet getRows(String baseQuery,
                           List<RowSetResource.RowSetRequest.RowRequestPair> preds,
                           int limit,
                           int offset) throws SQLException {
+        // TODO do we need to handle time column here?
```
What are you worried about?

viveksjain replied:
Do we need to order by the time column here as well? Also, the approach in this file seems rather hacky in general - I'm currently assuming `baseQuery` doesn't already contain the `ORDER BY` etc etc. But I guess it's good enough for our purposes for now.

pbailis replied:
Yeah, `baseQuery` is a hack. As an intermediate option, you could check for `ORDER BY` and error out.

pbailis commented on src/main/java/macrobase/ingest/SQLLoader.java:
```diff
@@ -213,5 +233,19 @@ public RowSet getRows(String baseQuery,
         return ret;
     }
+    
+    // Shield your eyes, mere mortals, from this glorious hideousness.
```
:100:

viveksjain replied:
This is how I vent.

pbailis replied:
:+1:

pbailis commented:
This is looking good. My main two comments are:

 1. Let me know if this "pane" idea doesn't make sense. It's marginally more general and will allow us to have arbitrarily-sized window slides.
 2. I'm not a huge fan of having timeColumn all over the code. It may be the cleanest way in the end, but, nevertheless, have you thought about alternatives?

I liked how you handled the `OutlierDetector` inheritance.

viveksjain commented:
Re: 2. I don't know what the alternatives are. If you have other ideas, I'm all ears. The only thing I can think of is that `timeColumn, attributes, lowMetrics, highMetrics` are used together in a bunch of places - maybe it's worth refactoring them out into a single wrapper class.

viveksjain commented:
@pbailis good to merge for now? Main remaining TODOs: wrapper class for `timeColumn, attributes, lowMetrics, highMetrics`, tests, logging for numenta. For the last one, I can show you the log4j bug in person. Best approach is probably to add a flag; it would be nicer to do that after #76 so that I don't have to pass around the flag in a bunch of unnecessary places.

pbailis commented:
Can you add some basic tests? :100: 

Then :+1:

viveksjain commented:
Rebased and added basic tests. I decided to create DatumWithTime subclass - perhaps this is premature optimization, but this way we don't have to pay cost of the additional pointer in every Datum if the time isn't needed.

pbailis commented:
Great work! Can you squash for review?

viveksjain commented:
How come coveralls doesn't show `MovingAverage.java`?

pbailis commented:
Very strange, not sure...

Do you need me to take another look at this?

viveksjain commented:
Squashed!

pbailis commented:
Great. LMK if you need another look, otherwise feel free to merge.

viveksjain commented:
The server seems to still work (although of course, the web UI doesn't yet support time column) and the output matches output from master. So I'm going to be bold and
![](http://www.tshirtlaundry.com/assets/images/photos/GopherIt11-30-2010-2.jpg)

deepakn94 commented:
Ur comment game is strong.

pbailis commented:
:+1: :100:
